source/extensions/filters/http/admission_control/thread_local_controller.h

```diff
@@ -29,6 +29,79 @@ class ThreadLocalController {
   virtual uint32_t requestSuccessCount() PURE;
 };
 
+/**
+ * Thread-local object to track request counts and successes over a rolling time window. Request
+ * data for the time window is kept recent via a circular buffer that phases out old request/success
+ * counts when recording new samples.
+ *
+ * This controller is thread-local so that we do not need to take any locks on the sample histories
+ * to update them, at the cost of decreasing the number of samples.
+ *
+ * The look-back window for request samples is accurate up to a hard-coded 1-second granularity.
+ * TODO (tonya11en): Allow the granularity to be configurable.
+ */
+class ThreadLocalControllerImpl : public ThreadLocalController,
+                                  public ThreadLocal::ThreadLocalObject {
+public:
+  ThreadLocalControllerImpl(TimeSource& time_source, std::chrono::seconds sampling_window);
+  ~ThreadLocalControllerImpl() override = default;
+  void recordSuccess() override { recordRequest(true); }
+  void recordFailure() override { recordRequest(false); }
+
+  uint32_t requestTotalCount() override {
+    maybeUpdateHistoricalData();
+    return global_data_.requests;
+  }
+  uint32_t requestSuccessCount() override {
+    maybeUpdateHistoricalData();
+    return global_data_.successes;
+  }
```

Comment:
Instead of doing the maybe update twice, it seems like it would be easier to just have a function that returns the pair since I think they will almost always be accessed together?